adapter/internal/operator/apis/dp/v1alpha1/api_webhook.go

```diff
@@ -0,0 +1,115 @@
+/*
+ *  Copyright (c) 2022, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package v1alpha1
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/wso2/apk/adapter/internal/loggers"
+	"github.com/wso2/apk/adapter/pkg/logging"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+var c client.Client
+
+// SetupWebhookWithManager creates a new webhook builder for API
+func (r *API) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	c = mgr.GetClient()
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// TODO(user): EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+
+//+kubebuilder:webhook:path=/mutate-dp-wso2-com-v1alpha1-api,mutating=true,failurePolicy=fail,sideEffects=None,groups=dp.wso2.com,resources=apis,verbs=create;update,versions=v1alpha1,name=mapi.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Defaulter = &API{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *API) Default() {
+	// TODO: Add any defaulting logic here
+}
+
+// TODO(user): change verbs to "verbs=create;update;delete" if you want to enable deletion validation.
+//+kubebuilder:webhook:path=/validate-dp-wso2-com-v1alpha1-api,mutating=false,failurePolicy=fail,sideEffects=None,groups=dp.wso2.com,resources=apis,verbs=create;update,versions=v1alpha1,name=vapi.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Validator = &API{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *API) ValidateCreate() error {
+	loggers.LoggerAPKOperator.Infof("Validate API create: %s", r.Name)
+	return r.validateAPI()
```

Comment:
What happens if the validation fails? CR creation would get rejected or get created anyway?

Reply:
yes, resource doesn't get created in etcd as k8s server rejects with the error.  User can see the error in the console.